ethdb/delete_range.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package ethdb defines the interfaces for an Ethereum data store.
+package ethdb
+
+import (
+	"bytes"
+	"errors"
+)
+
+var ErrTooManyKeys = errors.New("too many keys in deleted range")
+
+// DeleteRangeWithIterator is a fallback method for deleting a key range from a
+// database that does not natively support range deletion.
+// Note that the number of deleted keys is limited in order to avoid blocking for
+// a very long time. ErrTooManyKeys is returned if the range has only been
+// partially deleted. In this case the caller can repeat the call until it
+// finally succeeds.
+func DeleteRangeWithIterator(db KeyValueStore, start, end []byte) error {
+	batch := db.NewBatch()
+	if err := deleteRangeWithIterator(batch, db, start, end); err != nil {
+		return err
+	}
+	return batch.Write()
+}
+
+// DeleteRangeFromBatch is a fallback method for deleting a key range in a batch
+// from a database that does not natively support range deletion.
+func DeleteRangeFromBatch(target Batch, source Iteratee, start, end []byte) error {
+	deleteRangeWithIterator(target, source, start, end)
```

Comment:
This immediately deletes the items, doesn't it? If so, that means that invoking `mybatch.DeleteRange(...)` will start deleting stuff immediately, and later on, when the batch is actually written it will do some post-processing to delete any delete-rangy values that it just wrote. 

If that's an accurate description, then that is not ok. A batch must not write any changes prior to it being committed. 

As I see it, the semantics of batched `DeleteRange` are pretty tricky to simulate without native support. Does it need to exist on the Batch interface, or can we get by with having it only on the database?

Reply:
Ah no, wait, my bad. So this just iterates the current state of the database, and adds the keys ot the batch. But then I'm not sure what the post-processing is all about?

Reply:
To delete any items previously added by the same batch. The iterator does not see those yet but I guess we should also remove those. Like if I do `batch.Put("123", "xxx")` and then `batch.DeleteRange("100", "200")` (assuming that there was no entry at "123" in the db before, only in the current batch).

Reply:
Maybe I should add a comment about this.

Reply:
Ah, right.
However,  it's still not correct. 
```
batch.DeleteRange("100","200)
db.Write("101", "foo")
batch.Write()
```
In order to properly simulate deleterange, the evaluation of keys to delete should happen later. Do we need it on batch?

Reply:
You are right. I do not need it on batch right now so I'm fine with not trying to emulate it. It's still a part of `KeyValueWriter` though so `Batch` should also implement it, at least with a dummy function that always errors. What I'm not sure about is whether we should support it on pebble and error on leveldb, or just always error. The argument for the second option is that someone might actually use it and not notice that it will not work on leveldb.

Reply:
Removed batch support (even from the pebble implementation), now `batch.DeleteRange` always panics.